src/main/scala/com/lightbend/kafka/scala/streams/ProcessorS.scala

```diff
@@ -0,0 +1,17 @@
+/**
+* Copyright (C) 2018 Lightbend Inc. <https://www.lightbend.com>
+*/
+
+package com.lightbend.kafka.scala.streams
+
+import org.apache.kafka.streams.processor.{AbstractProcessor, Processor, ProcessorSupplier}
+
+/**
+  * When adding a processor to Topology, what is added is not a processor itself, but rather a processor supplier
+  * This little abstract class hides this level of indirection and provides a default implementation of
+  */
+
+abstract class ProcessorS[K,V] extends AbstractProcessor[K,V] with ProcessorSupplier[K,V]{
```

Comment:
let me have my concerns here against going for an extra level of indirection .. The class `ProcessorSupplier` has only one method `get`, which means it qualifies as a SAM (Single Abstract Method) type. This is now completely handled as part of Scala.

Let's see how this changes things .. Assume we don't have the `ProcessorS` abstraction and we work with `AbstractProcessor`. The we have ..

```scala
class SampleProcessor(punctuateTime: Long) extends AbstractProcessor[String, String] {
  //..
}
```

Now where we add the processor we can do the following ..

```scala
topology.addProcessor("data processor", () => new SampleProcessor(5000), "data")
```

Just by making it a function literal, scalac will convert it to the SAM type of Java lambda. We don't need to pass the `ProcessorSupplier` explicitly. All of this will be taken care of by the underlying compiler infrastructure which supports SAM type.

What we can do just by passing a function literal I don't think we need to create an extra level of class hierarchy.

WDYAT ? @seglo, @blublinsky, @deanwampler ?

Reply:
@debasishg A few questions

1. The SAM type conversion using a function literal is fine, but may not be intuitive to end users unless they're aware of this Scala compiler feature.  [I just read up about this feature here](http://www.scala-lang.org/news/2.12.0/#lambda-syntax-for-sam-types)
2. What would happen if this technique with Scala 2.11.  AFAICT the SAM type conversation feature is only available in 2.12.